test/TestInternalGrains/StreamingImmutabilityTestGrain.cs

```diff
@@ -0,0 +1,56 @@
+﻿using System;
+using System.Threading.Tasks;
+using Orleans;
+using Orleans.Streams;
+using UnitTests.GrainInterfaces;
+
+namespace UnitTests.Grains
+{
+    public class StreamingImmutabilityTestGrain : Grain, IStreamingImmutabilityTestGrain
+    {
+        private StreamImmutabilityTestObject _myObject;
+        private StreamSubscriptionHandle<StreamImmutabilityTestObject> _streamSubscriptionHandle;
+
+        public async Task SubscribeToStream(Guid guid, string providerName)
+        {
+            var stream = GetStreamProvider(providerName).GetStream<StreamImmutabilityTestObject>(guid, "Namespace");
+            _streamSubscriptionHandle = await stream.SubscribeAsync(OnNextAsync);
+        }
+
+        public async Task UnsubscribeFromStream()
+        {
+            if (_streamSubscriptionHandle != null)
+                await _streamSubscriptionHandle.UnsubscribeAsync();
+        }
+
+        public async Task ProduceTestObject(string providerName)
+        {
+            _myObject = new StreamImmutabilityTestObject() {MyString = this.GetPrimaryKey().ToString()};
+            var stream = GetStreamProvider(providerName).GetStream<StreamImmutabilityTestObject>(this.GetPrimaryKey(), "Namespace");
+            await stream.OnNextAsync(_myObject);
```

Comment:
Immutable means that the application level promises not to modify the data, so the stream system can optimize around this promise and avoid data copies (faster).  To test if this is happening, you'd need the application level to break this promise by modifying the data and see if the sent data is similarly modified.

All that is tested here is that the object was send/received, not whether it was immutable or not.  To test that the immutable setting was applied, you must violate the implicit promise not to mutate the state.  Something like:
Task sendTask = streamOnNextAsync(_myObject)
_myObject.MyString = "blarg"
return sendTask.
Then in onNextAsync check to see if the received object's string is this.GetPrimaryKey().ToString() or "Blarg".   If it is "Blarg", then the stream was set to immutable, because the original state of the object when the send was made was not copied.